File: src/compiler/typescript_to_rust/generators/ic_object/functions/trap.ts
import { Rust } from '../../../../../types';

export function generateIcObjectFunctionTrap(): Rust {
    return /* rust */ `
        fn _azle_ic_trap(
            _this: &boa_engine::JsValue,
            _aargs: &[boa_engine::JsValue],
            _context: &mut boa_engine::Context
        ) -> boa_engine::JsResult<boa_engine::JsValue> {
            ic_cdk::api::trap(_aargs.get(0).unwrap().as_string().unwrap());
        }
    `;
}

File: src/compiler/typescript_to_rust/azle_generate/src/body/ic_object/functions/trap.rs
pub fn generate() -> proc_macro2::TokenStream {
    quote::quote! {
        fn trap(
            _this: &boa_engine::JsValue,
            aargs: &[boa_engine::JsValue],
            context: &mut boa_engine::Context,
        ) -> boa_engine::JsResult<boa_engine::JsValue> {
            let message: String = aargs
                .get(0)
                .ok_or_else(|| "An argument for 'message' was not provided".to_js_error(None))?
                .clone()
                .try_from_vm_value(context)?;

            ic_cdk::api::trap(&message);
        }
    }
}
